### PR TITLE
Integrate latest ClientWebSocket updates and fix to include error details

### PR DESCRIPTION
**Pull request by dlstucki**
## Description
- Update ClientWebSocket to the latest version in dotnet/corefx master*
- *Do some reflection magic to access SocketsHttpHandler and SslClientAuthenticationOptions. These are publicly available from .NET Core 2.1 and later (.NET Core 2.0 is "end of life").
- Fix updated ClientWebSocket to provide HttpResponseMessage for HTTP error details.

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [x] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module).
- [x] If applicable, the public code is properly documented.
- [x] Pull request includes test coverage for the included changes.
- [x] The code builds without any errors.

### CODE REVIEW

File: src/Microsoft.Azure.Relay/Strings.resx [hunk 1]
```diff
@@ -315,4 +315,10 @@
   <data name="ResponseBodyStarted" xml:space="preserve">
     <value>The HTTP status code, status description, and headers cannot be modified after writing to the output stream.</value>
   </data>
+  <data name="net_invalid_enum" xml:space="preserve">
+    <value>TODO</value>
+  </data>
+  <data name="net_WebSockets_MissingResponseHeader" xml:space="preserve">
+    <value>TODO</value>
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/Strings.resx]: Fix this.

File: src/Microsoft.Azure.Relay/WebSockets/NetCore/WebSocketHandle.Managed.cs [hunk 2]
```diff
@@ -77,503 +59,211 @@ public void Abort()
...
-                if (options.RequestedSubProtocols.Count > 0)
-                {
-                    builder.Append(HttpKnownHeaderNames.SecWebSocketProtocol).Append(": ");
-                    builder.Append(options.RequestedSubProtocols[0]);
-                    for (int i = 1; i < options.RequestedSubProtocols.Count; i++)
+                // Get the response stream and wrap it in a web socket.
+                Stream connectedStream = await response.Content.ReadAsStreamAsync().ConfigureAwait(false);
+                Debug.Assert(connectedStream.CanWrite);
+                Debug.Assert(connectedStream.CanRead);
+                MethodInfo createFromStreamMethod = typeof(WebSocket).GetMethod(
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/WebSockets/NetCore/WebSocketHandle.Managed.cs]: >createFromStreamMethod [](start = 27, length = 22)

Make sure this gets checked as non-null.

File: src/Microsoft.Azure.Relay/WebSockets/NetCore/WebSocketHandle.Managed.cs [hunk 3]
```diff
@@ -594,205 +284,23 @@ private static byte[] BuildProxyHeaders(Uri uri, ClientWebSocketOptions options,
...
-            // immediately closes the connection may either result in an exception getting thrown from the connect
-            // earlier, or it may result in getting to here but reading 0 bytes.  If we read 0 bytes and thus have
-            // an empty status line, treat it as a connect failure.
-            if (string.IsNullOrEmpty(statusLine))
+            if (!response.Headers.TryGetValues(name, out IEnumerable<string> values))
             {
-                // "The response status line value '' is invalid."
-                throw new WebSocketException(WebSocketError.HeaderError, SR.Format(SR.net_WebSockets_ResponseStatusInvalid, string.Empty));
+                var wsException = new WebSocketException(WebSocketError.Faulted, SR.Format(SR.net_WebSockets_MissingResponseHeader, name));
+                wsException.Data[typeof(HttpResponseMessage).FullName] = response;
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/WebSockets/NetCore/WebSocketHandle.Managed.cs]: >wsException.Data[typeof(HttpResponseMessage).FullName]  [](start = 16, length = 55)

This is how I preserve the HTTP error info.

File: src/Microsoft.Azure.Relay/WebSockets/NetCore/WebSocketHandle.Managed.cs [hunk 4]
```diff
@@ -594,205 +284,23 @@ private static byte[] BuildProxyHeaders(Uri uri, ClientWebSocketOptions options,
...
-            if (!foundUpgrade || !foundConnection || !foundSecWebSocketAccept)
+            Debug.Assert(values is string[]);
+            string[] array = (string[])values;
+            if (array.Length != 1 || !string.Equals(array[0], expectedValue, StringComparison.OrdinalIgnoreCase))
             {
-                // The response header 'Upgrade|Connection|Sec-WebSocket-Accept' was not found.
-                string missingHeaderName = !foundUpgrade ? HttpKnownHeaderNames.Upgrade : !foundConnection ? HttpKnownHeaderNames.Connection : HttpKnownHeaderNames.SecWebSocketAccept;
-                throw new WebSocketException(WebSocketError.HeaderError, SR.Format(SR.net_WebSockets_ResponseHeaderMissing, missingHeaderName));
+                var wsException = new WebSocketException(WebSocketError.HeaderError, SR.Format(SR.net_WebSockets_InvalidResponseHeader, name, string.Join(", ", array)));
+                wsException.Data[typeof(HttpResponseMessage).FullName] = response;
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/WebSockets/NetCore/WebSocketHandle.Managed.cs]: >wsException.Data[typeof(HttpResponseMessage).FullName] [](start = 16, length = 54)

This is how I preserve the HTTP error info.

File: src/Microsoft.Azure.Relay/WebSockets/NetCore/WebSocketHandle.Managed.cs [hunk 5]
```diff
@@ -77,503 +59,211 @@ public void Abort()
...
+                    response = await handler.SendAsync(request, externalAndAbortCancellation.Token).ConfigureAwait(false);
+                    externalAndAbortCancellation.Token.ThrowIfCancellationRequested(); // poll in case sends/receives in request/response didn't observe cancellation
                 }
-                else
+
+                if (response.StatusCode != HttpStatusCode.SwitchingProtocols)
                 {
-                    builder.Append(hostHeader).Append("\r\n");
+                    var wsException = new WebSocketException(WebSocketError.NotAWebSocket, SR.Format(SR.net_WebSockets_Connect101Expected, (int)response.StatusCode));
+                    wsException.Data[typeof(HttpResponseMessage).FullName] = response;
```
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/WebSockets/NetCore/WebSocketHandle.Managed.cs]: >wsException.Data[typeof(HttpResponseMessage).FullName] [](start = 20, length = 54)

This is how I preserve the HTTP error info.

File: src/Microsoft.Azure.Relay/WebSockets/NetCore/WebSocketHandle.Managed.cs (hunk 2 above)
Comment (dlstucki) [inline src/Microsoft.Azure.Relay/WebSockets/NetCore/WebSocketHandle.Managed.cs]: >WebSocket [](start = 59, length = 9)

Add a similar comment about how this is supported in netcoreapp2.1 but is not part of netstandard2.0.

File: src/Microsoft.Azure.Relay/WebSockets/NetCore21/ClientWebSocket.cs [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+            ThrowIfNotConnected();
+            return _innerWebSocket.ReceiveAsync(buffer, cancellationToken);
+        }
+
+        // Requires .Net Standard 2.1
+        ////public override ValueTask<ValueWebSocketReceiveResult> ReceiveAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+        ////{
+        ////    ThrowIfNotConnected();
+        ////    return _innerWebSocket.ReceiveAsync(buffer, cancellationToken);
+        ////}
```
Comment (bainian12345) [inline src/Microsoft.Azure.Relay/WebSockets/NetCore21/ClientWebSocket.cs]: Can maybe uncomment these here?

Comment (dlstucki) [reply src/Microsoft.Azure.Relay/WebSockets/NetCore21/ClientWebSocket.cs]: I don't feel that I can. These in turn call down to WebSocket methods which are only available in .NET Standard 2.1 Preview.

---
In reply to: [290883715](https://github.com/Azure/azure-relay-dotnet/pull/109#discussion_r290883715) [](ancestors = 290883715)

**Pull request merged**